src/Engine/ProtoCore/DSASM/HeapObjects.cs

```diff
@@ -0,0 +1,659 @@
+                {
+                    return false;
+                }
+
+                if (!StackUtils.CompareStackValues(value1, value2, rtCore1, rtCore2))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+    }
+
+    public class DSObject : HeapElement
+    {
+        public DSObject(int size, Heap heap)
+            : base(size, heap)
+        {
+            MetaData = new MetaData { type = (int)PrimitiveType.kTypePointer };
+        }
+
+        public StackValue GetValueFromIndex(int index, RuntimeCore runtimeCore)
+        {
+            if (index >= VisibleSize || index < 0)
+            {
+                if (runtimeCore != null)
+                    runtimeCore.RuntimeStatus.LogWarning(WarningID.kOverIndexing, Resources.kArrayOverIndexed);
+                return StackValue.Null;
+            }
+
+            return GetItemAt(index);
+        }
+
+        public StackValue SetValueAtIndex(int index, StackValue value, RuntimeCore runtimeCore)
+        {
+            if (index >= VisibleSize || index < 0)
+            {
+                if (runtimeCore != null)
+                    runtimeCore.RuntimeStatus.LogWarning(WarningID.kOverIndexing, Resources.kArrayOverIndexed);
+                return StackValue.Null;
+            }
+
+            StackValue oldValue = GetItemAt(index);
+            SetItemAt(index, value);
+            return oldValue;
+        }
+    }
+
+    public class DSString : HeapElement
+    {
+        public StackValue Pointer = StackValue.Null;
+
+        public DSString(int size, Heap heap)
+            : base(size, heap)
+        {
+            MetaData = new MetaData { type = (int)PrimitiveType.kTypeString };
+        }
+
+        public void SetPointer(StackValue pointer)
+        {
+            this.Pointer = pointer;
+        }
+
+        public string Value
+        {
+            get
+            {
+                return heap.GetString(this);
+            }
+        }
+
+        public StackValue GetValueAtIndex(int index, RuntimeCore runtimeCore)
+        {
+            string str = heap.GetString(this);
+            if (str == null)
+                return StackValue.Null;
+
+            if (index < 0)
+            {
+                index = index + str.Length;
+            }
+
+            if (index >= str.Length || index < 0)
+            {
+                if (runtimeCore != null)
+                    runtimeCore.RuntimeStatus.LogWarning(WarningID.kOverIndexing, Resources.kStringOverIndexed);
+                return StackValue.Null;
+            }
+
+            return heap.AllocateString(str.Substring(index, 1));
+        }
+
+        public StackValue GetValueAtIndex(StackValue index, RuntimeCore runtimeCore)
+        {
+            int pos = Constants.kInvalidIndex;
+            if (index.IsNumeric)
+            {
+                pos = (int)index.ToInteger().opdata;
```

Comment:
As I browse through the code, nearly 99% of the callers of `StackValue.ToInteger()` are doing this:

```csharp
var foo = sv.ToInteger().opdata;
```

They are all interested in getting its integer value, some would call `RawIntValue` or `RawDoubleValue`. I'm wondering if operators would have made the callers' intentions clearer:

```csharp
var integer = (int)sv; // Throws exception if it's not 'int'
var dbl = (double)sv; // Throws exception if it's not 'double'
var value = sv.ToPrimitive(); // either an 'int' or a 'double'
```

The last one can simply a lot of caller such as this one:

```csharp
// Existing way
var value1 = opdata2.IsDouble ? opdata2.RawDoubleValue : opdata2.RawIntValue;

// New way
var value1 = opdata2.ToPrimitive();
```

Perhaps another task, but it'll be interesting to see.